### ISSUE TITLE
Refactor code to facilitate VSMac support

### ISSUE THREAD

**Issue by corliss** [VsVim/VsVim#2655]
Hi! This request is related to https://github.com/nosami/XSVim/issues/252. More details below.

**Comment by corliss** [comment]
Based on information you provided @nosami, I did a little research. 

To summarize:

Visual Studio for Mac has moved greatly towards adopting the Visual Studio for Windows editor and extensibility interfaces:

https://docs.microsoft.com/en-us/visualstudio/releasenotes/vs2019-mac-preview-relnotes#preview3

https://docs.microsoft.com/en-us/visualstudio/mac/extending-visual-studio-mac?view=vsmac-2019#extending-the-new-editor

https://github.com/mono/monodevelop/wiki/New-Editor-API-Guide

Because VSVim's implementation uses the same VS interfaces that VS4Mac now supports, it seems best to extend VSVim to support VS4Mac. Additionally, VSVim is very well factored for such an approach.

However, since VS4Mac has a Cocoa based UI layer, in contrast to the Wpf layer in VS, as it currently stands a few refactorings in VSVim are needed.

The current dependency tree is:

```
VSVim
+ VimCore
+ VimWpf
  + VimCore
+ VsInterfaces
  + VimCore
  + VimWpf
+ VsVimShared
  + VimCore
  + VimWpf
  + VsInterfaces
```

`VimWpf` is a dependency of `VsInterfaces`, `VsVimShared`, and `VsVim`.
It would be great if the Wpf-specific code was abstracted out, possibly using MEF to be consistent with the rest of the project. Looking at the `VimWpf` code, it looks like there are several UI-agnostic interfaces and type definitions already, so I'm guessing this refactoring won't be very difficult (in theory at least.)

What do you think?
Thanks a lot.

**Comment by jaredpar** [comment]
I'm very supportive of making changes to enable VsVim to run in Vs4Mac. It would be really nice to have it working there. 

I'm curious though why you think that VsInterfaces will be necessary in Vs4Mac. The original purpose of VsInterfaces was to allow VsVim access to services specific to the version of Visual Studio it's currently being run in. For every version of Visual Studio that VsVim supports there is a helper DLL specific to that version. The VsInterfaces layer is simple abstraction over that layer. 

The Visual Studio specific DLLs are mostly specific to Visual Studio though. It's abstracting over DLLs like Microsoft.VisualStudio.Platform.WindowManagement which aren't officially supported or versioned and require runtime tricks to access safely. 

How would this be applicable to VS4Mac? Does it have a similar support story where it's common to have more than the "latest" version used by a significant number of users and APIs that are specific to versions of VS4Mac that would also be used in VsVim?

The one aspect that I can see being interesting to VS4Mac is the word completion support (<kbd>Contrl-N</kbd>). That today exists in the VsInterfaces layer but that's just because the VsInterfaces layer was a convenient way to ship the completion support (supporting legacy and async completion simultaneously is tricky). 

Depending on how VS4Mac servicing works there may be a simpler solution here than extending VsInterfaces. If not though then I think we could likely get through the refactoring you described 

Note: my responses will be delayed for the next few days as I'm currently moving.

**Comment by corliss** [comment]
Glad you are supportive of this. Also, the code seems very well written and maintained, well done!

Agree re: VsInterfaces, the Vs4Mac implementation will not need anything from it, apart from (eventually) the word completion support you mention. Since there is no backward compatibility story for Vs4Mac at this point (the new editor just came out of preview), there's no need for a similar abstraction at this time. I had just mentioned it because it depends on VimWpf.

Since we're agreed on this, will wait to hear back from you when this step is done, and then will resume the investigation/implementation of the port.

Have a safe move 😀

**Comment by nosami** [comment]
> How would this be applicable to VS4Mac? Does it have a similar support story where it's common to have more than the "latest" version used by a significant number of users and APIs that are specific to versions of VS4Mac that would also be used in VsVim?

We generally have a Stable version and a Preview version of VSMac out at any given time. The way that I handled this in XSVim was to build two versions of the Addin. There are separate Addin feeds for each version.

APIs may break between major versions but are usually binary compatible between minor versions.

**Comment by jaredpar** [comment]
Finally done with the move and back online a bit. 

> Since we're agreed on this, will wait to hear back from you when this step is done, and then will resume the investigation/implementation of the port.

I think the next step is actually to move forward with the Vs4Mac extension. Essentially get a VsMac (whatever name you prefer) project added, begin an implementation of `IVimHost` and getting the basics up and running in the Vs4Mac scenario. If you think it make sense then maybe have two projects:  VimCocoa project, logical peer to VimWpf, and VsMac project that references it. 

Once the basics are up and running I can get the word completion support working there. I think it should actually just be as easy as referencing the shared project directly and creating a new `#if VS4MAC` or similar. Should be pretty straight forward once I get a better feel for how VsMac project is setup.

**Comment by corliss** [comment]
The difficulty is that code reuse is not currently possible as explained below.

As mentioned earlier, VS4Mac supports the very same interfaces - `ITextBuffer` and `ITextView` - that are the basis of the `VSVimShared` assembly. This (quite large) assembly can, I believe, be reused in its entirety with one caveat - its tight coupling with `VimWpf`, which is pervasive. This prevents even building the assembly for Mac. What I'm requesting is for the tight coupling to be abstracted out so that the assembly builds on Mac as a first step. This refactoring does not require a Mac - it is enough to define an abstract set of interfaces in another assembly - say `VimUI`- where `VimWpf` provides a late-bound implementation of `VimUI`. `VimUI` should not depend on anything in `System.Windows`.  

Once `VsVimShared` depends on `VimUI` (and not directly on `VimWpf`) the next step can begin, namely the development of `VimCocoa`, the Mac implementation of `VimUI`.

Additional opportunities for reuse will likely present themselves after this first step. The goal is to minimize duplication between VsVim and VsMacVim.

Here are some posts by another team that successfully went this route:
https://blog.neteril.org/blog/2019/06/15/vsmac-android-editor-extensions/
https://blog.neteril.org/blog/2019/04/11/cross-platform-testing-visual-studio-editor-extensions/

A quote from the first post:

> Finally, as part of porting the editor to Cocoa, a number of public interfaces had to be de-WPF-ized and new types introduced to replace them. 

That's also what is being proposed here. This approach also allows good separation between the core - of which you are the expert - and the incremental development that can be done independently for VSMac once this refactoring is done. One alternative - of `#if`'s littered in the VimWpf and other assemblies, is not a fun prospect.

This refactoring should be fairly mechanical, and will provide a solid foundation for the cross-platform development.

 😊

**Comment by jaredpar** [comment]
> As mentioned earlier, VS4Mac supports the very same interfaces - ITextBuffer and ITextView - that are the basis of the VSVimShared assembly. This (quite large) assembly can, I believe, be reused in its entirety with one caveat 

The VsVimShared assembly is really entirely Visual Studio Shell specific. Looking through the references you will see items like:

- Microsoft.VisualStudio.Shell.Interop.10.0
- Microsoft.VisualStudio.Shell.Interop.11.0
- Microsoft.VisualStuido.Shell.14.0

I can't really imagine the VsVimShared project working on VS4Mac because those references won't work there. That's okay though because this is a very Visual Studio shell focused project. 

Stepping back I think the project names are confusing things here. I definitely get the break down of the editor APIs and the layer that the VS4Mac uses / doesn't use. VsVim is already factored today such that it should be usable in VS4Mac without substantial changes. Here is how the projects fit into the various layers:

- Editor (non WPF): VimCore 
- Editor (WPF portions): VimWpf
- Visual Studio Shell specific: VsInterfaces, VsVimShared, Vs2015, VS2017, VS2019

95% of the Vim logic though is in VimCore. The `IVimHost` host interface exported from that assembly lets the host do things like open windows, close windows, move tabs, etc ... Implementing that for Vs4Mac should be the starting point for getting vim working.

> Once VsVimShared depends on VimUI (and not directly on VimWpf) the next step can begin, namely the development of VimCocoa, the Mac implementation of VimUI.

VimCocoa though is logically the peer of VimWPF in my mind at least. It's putting the Cocoa UI finishes on the core editor just as VimWPF puts the WPF finishes on the core editor.  The VimWPF project only depends on VimCore. Why do you feel VsVimShared needs to be ported first when VimWPF doesn't depend on it today?

> One alternative - of #if's littered in the VimWpf and other assemblies, is not a fun prospect.

There are no `#if` in the VimWpf project though. The only shipping projects which have `#if` are the VimSpecific and VsSpecific. VimSpecific definitely needs to be re-used within Vs4Mac in some fashion (it has word completion). VsSpecific though is Visual Studio Shell specific.

Feel like we're talking past each other on some parts here. Really want this to works and hoping this clears things up. If not perhaps we should find a time / place to chat more directly on this. Don't think we're very far away here. 😄

**Comment by corliss** [comment]
> The VsVimShared assembly is really entirely Visual Studio Shell specific.

VsVimShared also contains the entire implementation of the Visual Studio Editor implementation, does it not? That's what this request is about. Before the New Editor was introduced to VS4Mac, it would not have been feasible to reuse VSVim on Mac. But by separating out the layers, this now becomes possible.

Since there's a platform dependency on the VS Shell, this also needs to be factored out. But it looks like a large body of VsVimShared will be reusable. Some more research:

There are 14000 lines of C# code in VsVimShared. 

The longest file, `VsVimHost.cs` is 200 lines long. It only produces a handful of compiler errors on the Mac. Some micro-examination might help here, so let's consider one method:

        /// <summary>
        /// Open up a new document window with the specified file
        /// </summary>
        public override FSharpOption<ITextView> LoadFileIntoNewWindow(string filePath, FSharpOption<int> line, FSharpOption<int> column)
        {
            try
            {
                // Open the document in a window.
                VsShellUtilities.OpenDocument(_vsAdapter.ServiceProvider, filePath, VSConstants.LOGVIEWID_Primary,
                    out IVsUIHierarchy hierarchy, out uint itemID, out IVsWindowFrame windowFrame);

                // Get the VS text view for the window.
                var vsTextView = VsShellUtilities.GetTextView(windowFrame);

                // Get the WPF text view for the VS text view.
                var wpfTextView = _editorAdaptersFactoryService.GetWpfTextView(vsTextView);

                if (line.IsSome())
                {
                    // Move the caret to its initial position.
                    var snapshotLine = wpfTextView.TextSnapshot.GetLineFromLineNumber(line.Value);
                    var point = snapshotLine.Start;
                    if (column.IsSome())
                    {
                        point = point.Add(column.Value);
                        wpfTextView.Caret.MoveTo(point);
                    }
                    else
                    {
                        // Default column implies moving to the first non-blank.
                        wpfTextView.Caret.MoveTo(point);
                        var editorOperations = EditorOperationsFactoryService.GetEditorOperations(wpfTextView);
                        editorOperations.MoveToStartOfLineAfterWhiteSpace(false);
                    }
                }

                return FSharpOption.Create<ITextView>(wpfTextView);
            }
            catch (Exception e)
            {
                _vim.ActiveStatusUtil.OnError(e.Message);
                return FSharpOption<ITextView>.None;
            }
        }

There are a couple of calls to shell methods, to open the document and get the text view. These need to be made platform independent (will investigate further whether VS4Mac has something very similar, but hoisting these into a platform independent interface will work in any case.)

The rest of the method, even though it references an `IWpfTextView`, actually only uses the base `ITextView` defined methods. All the code related to getting a snapshot, getting a line from a line number, moving the caret, etc, will run as-is on Mac.

This kind of refactoring that would allow all the complex, error-prone logic to live in one place, and let the boring, platform details be implemented per-platform.

Also hope this is helping move towards a consensus :) Just FYI, this is something I've been researching on and off for a couple of months. Still feels like drinking from a firehose, what with learning more about Vim than I ever expected, the internals of two quite complex plugins, two different platform hosting architectures (old and new), and so on. A clean solution that integrates well with VsMac, and is fun to work on, is the hope.

**Comment by corliss** [comment]
Update: In the interest of progress, spent some time refactoring just the main file, VimHost.cs, moving the shell specific code to a new `iVsVimShell` interface. The result feels quite clean, and retains most of the plugin logic in `VimHost.cs`, which is what we want.

The work in progress is at https://github.com/corliss/VsVim/commit/87fed65dbef40c8da98fb3467804233f706cb505.

Let me know what you think when you get the chance. 
:)

**Comment by jaredpar** [comment]
> There are 14000 lines of C# code in VsVimShared.

Indeed but that's pretty much universally about supporting the Visual Studio shell environment. Visual Studio is a complex product with lots of legacy systems to deal with an a vibrant plugin ecosystem. All of which VsVim needs to interact with in some way.

I usually joke that 1/2 of the code in VsVim is implementing Vim on the editor and the other 1/2 is dealing with quirks of the Visual Studio Shell. 

Example: a good chunk of the code in VsVimShared is just about keyboard routing. Visual Studio has at least four different ways to route keyboard input and many sub-systems that hook it in various ways. VsVim being keenly interested in keyboard input has to interact with all of these systems, intercept keys at the appropriate time and forward them to the core Vim types which do the actual work. None of this though will be relevant to VS4Mac. 

> The longest file, VsVimHost.cs is 1200 lines long. It only produces a handful of compiler errors on the Mac

That is definitely the type which is most likely to have re-usable code for VS4Mac. Rather than break a part VsVimShared though I'd just push that code down into VimCore and refactor `IVimHost` to put the shared logic there.

Outside of that type though i don't imagine a lot of VsVimShared will be re-usable in VS4Mac. Are there other types there that you think are interesting? If there are I'd be very in favor of pushing them down to VimCore so they could be re-used.

> This kind of refactoring allows all the complex, error-prone logic to live in one place, and let the boring, platform details be implemented per-platform.

Agree. There is definitely code that lives in VsVimShared that could be pushed down to a lower layer, the bulk of that will be in `VsVimHost`. Part of the reason it's not cleanly factored at the moment is that there is only one real host to consider: Visual Studio. Hard to reason about what will and won't be host specific until you have more than one host to consider. VS4Mac gives a good opportunity to look at that. 

> The work in progress is at corliss@87fed65.

Looked at that a bit. What do you see as the difference between `IVimHost` and `IVsVimShell`? The intent of `IVimHost` was to be the host abstraction layer. Code which was host independent was meant to live inside `VimCore` and call out to `IVimHost` when necessary. `IVsVimShell` seems to be another host interface here. 

Looking at the concrete example you provided, which I agree isn't factored as well as it could be. The fix though in my mind is to change the implementation to be the following:

```
public override FSharpOption<ITextView> LoadFileIntoNewWindow(string filePath, FSharpOption<int> line, FSharpOption<int> column)
{
    try
    {
        // Open the document in a window.
        VsShellUtilities.OpenDocument(_vsAdapter.ServiceProvider, filePath, VSConstants.LOGVIEWID_Primary,
            out IVsUIHierarchy hierarchy, out uint itemID, out IVsWindowFrame windowFrame);

        // Get the VS text view for the window.
        var vsTextView = VsShellUtilities.GetTextView(windowFrame);

        // Get the WPF text view for the VS text view.
        var wpfTextView = _editorAdaptersFactoryService.GetWpfTextView(vsTextView);

        return FSharpOption.Create<ITextView>(wpfTextView);
    }
    catch (Exception e)
    {
        _vim.ActiveStatusUtil.OnError(e.Message);
        return FSharpOption<ITextView>.None;
    }
}
```

Then move all of the deleted code around cursor positioning into VimCore. and the code that calls `LoadFileIntoNewWindow`. There shouldn't need to be another interface here that is needed. 

Note: looking at the change it seems you're mixing tabs and spaces a bit.

**Comment by jaredpar** [comment]
Looking at `LoadFileIntoNewWindow` this is the approach I believe we should take for fixing it: https://github.com/VsVim/VsVim/commit/d4a0ad800a670a3fd8016ed79d50b1dc6ae05625 Essentially taking the host independent part and pushing it down to VimCore.

The reasoning behind this change is as follows:

VimCore is the project where host independent Vim behavior is intended to live. All of the host specific operations are expressed through `IVimHost` which every host needs to implement and export. VsVimShared is intended to be the Visual Studio specific implementation of `IVimHost`. 

The overwhelming majority of code in VsVimShared should be specific to Visual Studio. We've definitely slipped up a few times over the years and let host independent code end up there. That should be the exception though, not the rule. 

For that code though the preferred fix is just moving that code down to VimCore so it can be shared with all of the hosts. Refactoring VsVimShared to be host indepedent is a pretty big under taking and not what the original ideal was here. It was meant to be a Visual Studio specific host. 

Consider as an example VimApp. That is another implementation of `IVimHost` that does not reference VsVimShared. It has no Visual Studio dependency and hence doesn't need to. Yet it's a highly functional Vim editor. It's great for helping keep our host abstraction clean and a fast F5 loop. 

This seems like the same place VS4Mac should be at. Essentially one layer above VimCore. As we find features that are host indepndent but drifted into VimWpf, or VsVimShared, we should just refactor those down into VimCore which was the original intent.

**Comment by ricksladkey** [comment]
I personally wrote most of that code example, and as you noted, the accidental use of IWpfTextView is entirely unnecessary for the state semantics. In my defense, it is a bit of a nightmare trying to navigate the myriad of confusing APIs to find the right one. When it works, you are so grateful that your standards might suffer a little!

And Jared's proposal for rewriting is excellent, one that wasn't possible until we somehow managed to get a reference to the new text view. But since we **do** have a reference to the text view, and since we do return it to "safe land" VimCore, we are capable of doing much more than just caret positioning in the new window.

However, I think one of Jared's points is worth restating: VimApp is a perfectly usable editor and it has only a bare-bones implementation of IVimHost. So you can start with a very minimal IVimHost can ignore all of VsVimShared. There is very little critical logic in there. Most of it enables specialized features like `:tabe` or `:vsc`, etc.

Having said that, there is a fairly large amount of code that currently depends on WPF (and not accidentally) that **is** fairly critical, most notably the block caret and the command margin. The agnostic logic of these items is currently mixed with heavy use of WPF data types: `TextBox`, `Key`, events, etc. To just get it working, the block caret can be omitted, and the command margin can be greatly simplified.

In retrospect, the command margin could have been implemented as an `ITextView` and set up to process `KeyInput` values just line an ordinary buffer. If Vs4Mac supports that approach, it might be worth using it as a proof-of-concept and then adapting VsVim's command margin to use the new one.

**Comment by corliss** [comment]
(Was about to hit send when Rick's post came though, so I'll send anyway and read his post after.)

> Note: looking at the change it seems you're mixing tabs and spaces a bit.

:) Already noted previously. and fixed on my machine (thanks, .editorconfig!):

> p.s. Just noticed that I need to update my tabs-to-spaces setting to spaces for this project.



> Outside of that type though i don't imagine a lot of VsVimShared will be re-usable in VS4Mac. Are there other types there that you think are interesting? If there are I'd be very in favor of pushing them down to VimCore so they could be re-used.

You are the best person to answer that question :) I have tried to make the case that there is a fairly comprehensive refactoring to be done to minimize duplication between Vs and VsMac. I see VsMac as just another flavor of Vs, with a few more differences than otherwise. 

`IVimHost` is a well thought out interface that works with types that are already present on VsMac, and uses no WPF.  However, `VsVimShared` contains both view code and view model code. By 'view code' I mean both the WPF code and the VS Shell code. The refactoring should cleanly separate all the view code  elsewhere, but retain the view model code, introducing new view models where needed, so that all logic is kept in one place. Both Vs and VsMac are IDE's with equivalent functionality, and so the view models should cover all the needed functionality. In this way, `VsVimShared` can both build and be used on Mac, thus living up to its name :)

Having spent long hours in the last day looking at `VsVimShared` I conclude that there are thousands of lines that are needed and would work on Mac, too numerous (and pointless) to individually list and debate. At this point you know as much as I do about the requirements. Hope you have conducted (or will conduct) a similar exercise, and come to the same conclusion. 

In that case, we can then agree that such a refactoring needs to be done by you, the code owner, as you see fit, before work on VsMac can really get underway. That remains the purpose of this request.

**Comment by corliss** [comment]
This post is to start detailing some of the important functionality now available on VsMac. 
Note: these are what are available via the Assembly Browser, and have not been tested.

Please add to this list! :)

```
Microsoft.VisualStudio.Editor
    ITextMateService

Microsoft.VisualStudio.Text
    ITextView
    ITextBuffer
    ITextSnapshot
    CaretPosition
    SnapshotSpan
    
Roslyn
```

**Comment by corliss** [comment]
> In retrospect, the command margin could have been implemented as an ITextView and set up to process KeyInput values just line an ordinary buffer. If Vs4Mac supports that approach, it might be worth using it as a proof-of-concept and then adapting VsVim's command margin to use the new one.

Looking at the full `ITextView` definition, it sure seems like overkill, don't you think? Does it need all the snapshot, buffer graph, etc., capabilities? If not would a minimal view model definition suffice?

But you're right, that thought process that will hopefully drive the extraction of the needed view models.

**Comment by corliss** [comment]
Update: Am going ahead with the VsVimMac project, since the envisioned refactoring is clearly a big project. In the meantime the small refactorings (such as the one for `LoadFileIntoNewWindow` are welcome.) Let's see what happens.

To set expectations, this is actually a very busy time. Am using this as a kind of welcome distraction, so no promises...

I'm going to use a slightly different coding style, but things like spaces etc. will match up. Any strong preferences?

**Comment by jaredpar** [comment]
> You are the best person to answer that question :) 

I feel like I have answered this though and my words just aren't making traction here for whatever reason. There is some code in VsVimShared that could work in VS4Mac but it's a small part of the project. Also it's not a part of the core vim feature set. The core feature set of motions, modes, macros, key mapping, command line parsing, vimrc, etc ... are all in the VimCore project. An implementation of `IVimHost` that did next to nothing would get all of this without any changes to VsVimShared.

The VsVimShared project is more about dealing with host quirks. Some code that is host agnostic has slipped in here. The solution though is to move that code into VimCore not to refactor VsVimShared to have a host agnostic layer. That project was never intended to be host agnostic. The refactoring you're seeking to do here, create a host agnostic layer, has already been done: it's VimCore.

> I have tried to make the case that there is a fairly comprehensive refactoring to be done to minimize duplication between Vs and VsMac. I see VsMac as just another flavor of Vs, with a few more differences than otherwise.

That is true for the editor layer of VS and VS4Mac. 100% agree. The VsVimShared layer though is dealing with the Visual Studio Shell. I don't see how the vast majority of that code will have any bearing in VS4Mac. If you feel that's wrong can you provide specific examples? The ones I'm aware of are:

- Parts of `VsVimHost`
- Roslyn external editor plugin 
- NavigateTo hooks

The rest of VsVimShared is dealing with items like the VS keyboard binding settings, OLE layer, `IOleCommandTarget`, Visual Studio Option Pages, ReSharper integration, etc ... None of that will work in VS4Mac. 

> I conclude that there are thousands of lines that are needed and would work on Mac, too numerous (and pointless) to individually list and debate.

I'm not asking you to list the lines of code, I'm asking for feature areas, files, etc ... You're asking for a major refactoring of the code, one which is redundant with the current layering of VsVim, but label the idea of listing out the features needed as "pointless".  This level of refactoring needs a good justification particularly because the layering you want to achieve in the refactoring already exists. VimCore is where host agnostic code lives and any such code in VsVimShared should simply be moved there. 

I've taken the time to list the features I think can be shared between the hosts. It's a reasonable ask for you to do the same. 

> If that happens, we can then agree that such a refactoring needs to be done by you, the code owner, as you see fit, before work on VsMac can really get underway. That remains the purpose of this request.

I still feel such a refactoring is unnecessary. The goal of the refactoring is to create a host agnostic layer. That already exists. The features which are needed for VS4Mac should instead be ported from VsVimShared to VimCore. I'm more than happy to work on this for any areas you feel VS4Mac needs. I've listed a few areas I think could be shared. If you have more please list them out and we can work through them.

What would convince me such a refactoring was necessary is if there were reference assemblies which were shared between VS4Mac and Visual Studio but not a generally portable assembly. Basically a reference which couldn't say be reasonably loaded VimApp. At this point though I don't see any of those listed. The APIs shared between VS4Mac and VimApp is the core editor AFAIK. Hence code sharing between the two projects is best done at the VimCore level. 

> Am going ahead with the VsVimMac project, since the envisioned refactoring is clearly a big project

I strongly suggest you take the approach I recommend here. If you simply create a VimMac project, have it reference VimCore, get a minimal `IVimHost` implementation defined you can get up and running with Vim support in VS4Mac likely in just a few days. It really shouldn't take a lot of work to get this going and it will give you 90%+ of the features. 

Attempting to refactor VsVimShared first is going to take a **lot** of time and as I've stated several times I don't feel you've made the case that it's necessary. The major items you've listed so far as inclusion into VS4Mac can be done by moving them into VimCore or getting VimSpecific to work in VS4Mac. I'm more than happy to help move features to VimCore. Once there is a VimMac project to look at it will better help us to understand the architecture there and we should be able to pretty rapidly identify other features that can move into VimCore. 

VimSpecific is in the same boat. Once there is a VimMac project to work with I expect it will take me ~1-3 hours to get the completion APIs working there. 

My strong preference here is we focus on getting VsVim hosted in VS4Mac first. This should only take a few days for someone familiar with VS4Mac. Once it's there it will be fairly easy for us all to work on it in parallel. If in trying to move services from VsVimShared to VimCore we find that really there should be another layer which is above VimCore but below VsVimShared then we can add that project in. 

> I'm going to use a slightly different coding style, but things like spaces etc. will match up. Any strong preferences?

Please use the coding style already in use in this project. We require all code to match this style.

**Comment by corliss** [comment]
A quick reply to set you at ease:
> I strongly suggest you take the approach I recommend here.
Yes, that is what I meant. :)

**Comment by ricksladkey** [comment]
> Looking at the full `ITextView` definition, it sure seems like overkill, don't you think? Does it need all the snapshot, buffer graph, etc., capabilities? If not would a minimal view model definition suffice?

It might seem like overkill, but many simple windows in Visual Studio are text views:

- The output window
- The immediate window
- Interactive windows like C#, etc.
- All peek windows

It is really the right UI-agnostic tool for the job. Specifically, it abstracts all the problematic parts of WPF:

- Text contents
- Multiline support
- UTF32 code points
- Caret position
- Selection start and end
- Layout, visibility and scrolling
- Text formatting (color, font, and taggers)
- Extensibile for adornments like control characters

These are all things that the command margin needs that a text view does already without any dependence on WPF. Anyway, it is an approach worth considering.

**Comment by ricksladkey** [comment]
The last thing you will need to get a "first working version" is keyboard input. The provided extension point in Visual Studio is explicitly WPF-based (`KeyProcessor`). Likewise for mouse input, but dealing with mouse input can be delayed until later versions.

Rudimentary keyboard input is easy; you just have to convert UTF16 characters to key inputs, convert and apply key modifiers, and dispatch them to the correct vim buffer. Full keyboard input is messy with many traps, dead keys, IMEs, the Alt key, AltGr, etc.

I propose an evolutionary approach where you cut-and-paste any code you think you need from VsVimShared as a temporary stopgap and then notate it with `TODO: DRY - VsVimShared` or some such comment. Once it compiles and runs and has tests, then any neutral reusable code will be much easier to identify and refactor. At that point we can make changes with confidence: eliminate the repetition while continuing to pass all tests.

For example, there is no value in moving something like `VisualStudioCommandDispatcher` out of VsVimShared if no other platform implements `EnvDTE` and `IVsUIShell`. It's not that the command dispatcher depends on WPF, but it might depend on  APIs that are only available when WPF is also available.

Surely there will be some counterexamples, but concrete cases are easier to reason about.

**Comment by corliss** [comment]
Just had a yikes moment - will the plugin size be the same order of magnitude as VsVimShared + WimWpf + VsVim? That's around 25000 lines of code. By comparison, the entire sixth edition of Unix was around 9000 loc. Am used to writing quite pithy code myself - lines should count! And the 100 lines/day figure has proven pretty accurate in my experience. And that is on full-time projects!

Am going to have to rethink this one, folks. At the start of this request I was optimistic that a small, concise host would be the outcome. But this discussion doesn't give me much confidence that that will be the case. Might just continue using the old editor as long as possible, or switch to VS code, which I've had decent experiences with.

Will get back to you on after mulling this one. Thoughts appreciated.

**Comment by ricksladkey** [comment]
I would not get discouraged. I am certain that a working editor without all the bells and whistles of tight Visual Studio integration will only require about 2000 lines of code, some of which would be boilerplate, and most of which would be have obvious parallels to existing code. For a better comparison, just look at the top-level lines in VimApp and VimWpf.

**Comment by corliss** [comment]
In summary, feel that the timing may be premature. The VsMac team seems to be on the trajectory toward more complete compatibility with Visual Studio. If that's the case, it's probably best to wait rather than jump in with a moving target. There are as yet quite significant differences that would make developing a Mac version of VsVim quite an effort.

Thanks for this useful discussion!

**Comment by nosami** [comment]
@jaredpar @ricksladkey Thanks for all the useful information that you have compiled in this issue. I am going to take a stab at getting VSVim working in VSMac in some fashion and we can take it from there. 
I will start with an IVimHost implementation that does as little as I can get away with and concentrate on the plumbing code needed to get the extension working on VSMac. Once I get that far, hopefully we can collaborate on this. I would love to see VSVim working on the Mac and then I can stop supporting XSVim at some point in the future.

I know the VSMac addin system and IDE code, but not so familiar with the new editor, MEF or this extension but it doesn't look too difficult to at least get _something_ working.
